core/client/src/test/java/alluxio/client/netty/NettyRemoteBlockReaderTest.java

```diff
@@ -0,0 +1,189 @@
+    Whitebox.setInternalState(mNettyRemoteBlockReader, "mHandler", mClientHandler);
+
+    ChannelFuture channelFuture = Mockito.mock(ChannelFuture.class);
+    Channel channel = Mockito.mock(Channel.class);
+    Mockito.when(channel.close()).thenReturn(channelFuture);
+    Mockito.when(channelFuture.sync()).thenReturn(channelFuture);
+    Mockito.when(channelFuture.channel()).thenReturn(channel);
+    Mockito.when(mBootstrap.connect(Mockito.any(SocketAddress.class))).thenReturn(channelFuture);
+  }
+
+  /**
+   * Test case for a valid {@link RPCBlockReadResponse} is received.
+   */
+  @Test
+  public void readRemoteBlockTest() throws IOException {
+    Thread channelThread = new Thread(new ChannelReadThread(
+            RPCResponse.Type.RPC_BLOCK_READ_RESPONSE, RPCResponse.Status.SUCCESS
+    ));
+    channelThread.start();
+
+    ByteBuffer byteBuffer = mNettyRemoteBlockReader.readRemoteBlock(mInetSocketAddress,
+            mBlockId, mOffset, mLength, mLockId, mSessionId);
+
+    Assert.assertEquals(10, byteBuffer.capacity());
+    Assert.assertEquals(10, byteBuffer.limit());
+    byte[] dst = new byte[10];
+    byteBuffer.get(dst);
+    Assert.assertEquals("alluxio", new String(dst).substring(0, 7));
+  }
+
+  /**
+   * Test case for an invalid {@link RPCBlockReadResponse} is received.
+   */
+  @Test(expected = IOException.class)
+  public void readRemoteBlockWithBadStatusTest() throws IOException {
+    Thread channelThread = new Thread(new ChannelReadThread(
+            RPCResponse.Type.RPC_BLOCK_READ_RESPONSE, RPCResponse.Status.BLOCK_LOCK_ERROR
+    ));
+    channelThread.start();
+
+    mNettyRemoteBlockReader.readRemoteBlock(mInetSocketAddress,
+            mBlockId, mOffset, mLength, mLockId, mSessionId);
+  }
+
+  /**
+   * Test case for {@link RPCErrorResponse} is received.
+   */
+  @Test(expected = IOException.class)
+  public void readRemoteBlockErrorResponseTest() throws IOException {
+    Thread channelThread = new Thread(new ChannelReadThread(
+            RPCResponse.Type.RPC_ERROR_RESPONSE, RPCResponse.Status.SUCCESS
+    ));
+    channelThread.start();
+    mNettyRemoteBlockReader.readRemoteBlock(mInetSocketAddress,
+            mBlockId, mOffset, mLength, mLockId, mSessionId);
+  }
+
+  /**
+   * Test case for {@link RPCFileWriteResponse} is received.
+   */
+  @Test(expected = IOException.class)
+  public void readRemoteBlockUnknownResponseTest() throws IOException {
+    Thread channelThread = new Thread(new ChannelReadThread(
+            RPCResponse.Type.RPC_FILE_WRITE_REQUEST, RPCResponse.Status.SUCCESS
+    ));
+    channelThread.start();
+    mNettyRemoteBlockReader.readRemoteBlock(mInetSocketAddress,
+            mBlockId, mOffset, mLength, mLockId, mSessionId);
+  }
+
+  /**
+   * Helper Thread for calling
+   * {@link ClientHandler#channelRead0(ChannelHandlerContext, RPCMessage)}.
+   .   */
+  private class ChannelReadThread implements Runnable {
```

Comment:
I tried this and it seems to work:
```java
    Mockito.when(mChannel.writeAndFlush(Mockito.any())).then(new Answer<ChannelFuture>() {
      @Override
      public ChannelFuture answer(InvocationOnMock invocation) throws Throwable {
        mClientHandler.channelRead0(null, createRPCBlockReadResponse(RPCResponse.Status.UFS_READ_FAILED));
        return null;
      }
    });
```

Basically, we can trigger the channelRead0 in response to the the request being sent through `writeAndFlush`. I think we need to do it this way so that we don't have synchronization issues between the `ChannelReadThread` and the call to `readRemoteBlock`.